File: backend/app/schemas/payload.py
from typing import Optional, Literal
from datetime import datetime
from pydantic import BaseModel


# ── PayloadType ────────────────────────────────────────────────────────────────

class PayloadTypeCreate(BaseModel):
    name: str
    manufacturer: str
    model: str
    category: Literal["sensor", "combat", "comms", "other"] = "sensor"
    weight_kg: float = 0.0
    voltage_v: float = 5.0
    max_current_a: float = 2.0
    has_gimbal: bool = False
    sensor_type: Optional[str] = None
    resolution: Optional[str] = None
    frame_rate_fps: Optional[float] = None
    payload_function: Optional[str] = None
    effective_range_m: Optional[float] = None
    notes: Optional[str] = None


class PayloadTypeUpdate(BaseModel):
    name: Optional[str] = None
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    category: Optional[Literal["sensor", "combat", "comms", "other"]] = None
    weight_kg: Optional[float] = None
    voltage_v: Optional[float] = None
    max_current_a: Optional[float] = None
    has_gimbal: Optional[bool] = None
    sensor_type: Optional[str] = None
    resolution: Optional[str] = None
    frame_rate_fps: Optional[float] = None
    payload_function: Optional[str] = None
    effective_range_m: Optional[float] = None
    notes: Optional[str] = None


class PayloadTypeOut(BaseModel):
    id: int
    name: str
    manufacturer: str
    model: str
    category: str
    weight_kg: float
    voltage_v: float
    max_current_a: float
    has_gimbal: bool
    sensor_type: Optional[str] = None
    resolution: Optional[str] = None
    frame_rate_fps: Optional[float] = None
    payload_function: Optional[str] = None
    effective_range_m: Optional[float] = None
    notes: Optional[str] = None
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/payload.ts
import { api } from './client'

export interface PayloadType {
  id?: number
  name: string
  manufacturer: string
  model: string
  category: 'sensor' | 'combat' | 'comms' | 'other'
  weight_kg: number
  voltage_v: number
  max_current_a: number
  sensor_type?: string
  resolution?: string
  frame_rate_fps?: number
  has_gimbal: boolean
  payload_function?: string
  effective_range_m?: number
  notes?: string
  is_active?: boolean
}

export const payloadApi = {
  listTypes: () => api.get<PayloadType[]>('/api/master/payload-types'),
  createType: (payload: PayloadType) => api.post<PayloadType>('/api/master/payload-types', payload),
  updateType: (id: number, payload: PayloadType) =>
    api.put<PayloadType>(`/api/master/payload-types/${id}`, payload),
  deleteType: (id: number) => api.delete(`/api/master/payload-types/${id}`),
  assignToDrone: (droneId: number, payloadTypeId: number | null) =>
    api.post(`/api/master/drones/${droneId}/payload`, { payload_type_id: payloadTypeId }),
}
